### ISSUE TITLE
[BUG] balls_left computed from whole overs only — model trained with ball-level precision but inference always assumes ball=0 of each over, producing wrong RRR for every mid-over prediction

### ISSUE THREAD

**Issue by anshul23102** [Arnav-Singh-5080/IPL-Win-Probability-Prediction#155]
## Describe the Bug

The win probability model is trained with per-ball granularity but the prediction UI only accepts a whole-over input. This creates a persistent feature mismatch that produces systematically incorrect `balls_left` and `rrr` values for every prediction made mid-over.

### Training code (ball-level precision)

**File:** `application.py`

```python
# Training — uses both over AND ball within that over
df['balls_left'] = 120 - (df['over'] * 6 + df['ball'])
df['crr']        = df['current_score'] / (df['over'] + df['ball'] / 6)
df['rrr']        = (df['runs_left'] * 6) / df['balls_left']
```

### Prediction code (coarse over-level only)

**File:** `application.py`

```python
# Prediction — ball column is completely absent
balls_left = 120 - (overs * 6)       # always assumes ball = 0 of the over
crr        = score / overs            # always divides by whole overs
rrr        = (runs_left * 6) / balls_left if balls_left > 0 else 0
```

### Concrete example

A match is at **over 15, ball 3** (i.e., 15.3 overs). The true state:

| Feature | Training value | Prediction value | Error |
|---|---|---|---|
| `balls_left` | 120 - (15*6 + 3) = **57** | 120 - (15*6) = **60** | **+3 balls** |
| `crr` (score=120) | 120 / (15 + 3/6) = **7.74** | 120 / 15 = **8.00** | **+0.26** |
| `rrr` (runs_left=60) | (60*6) / 57 = **6.32** | (60*6) / 60 = **6.00** | **-0.32** |

The model is told there are 3 more balls remaining than there actually are. The effective run rates are wrong in opposite directions, compounding the prediction error. This happens on every single prediction unless the user happens to be exactly at the start of an over.

## Steps to Reproduce

1. Open the Analysis page
2. Set Overs to 10 (slider selects whole overs only — no sub-over field)
3. Run a prediction at e.g. 10 overs, 80 runs, target 160
4. The model receives `balls_left=60` but during training, ball 10.3 would give `balls_left=57`

## Expected Behavior

Predictions should accept a "balls bowled in current over" input (0-5) and derive:

```python
balls_left = 120 - (overs * 6 + balls_in_over)
crr        = score / (overs + balls_in_over / 6) if (overs + balls_in_over / 6) > 0 else 0
rrr        = (runs_left * 6) / balls_left if balls_left > 0 else 0
```

## Suggested Fix

Add a `st.slider("Balls in Current Over", min_value=0, max_value=5, value=0, step=1, key="balls_in_over")` widget alongside the overs input. Wire it into the `balls_left` and `crr` derivations to match training-time feature engineering exactly.

## Impact

Every win probability prediction made at a non-zero ball within an over is computed from incorrect feature values. Because the over slider forces integer overs, **all real-world mid-over queries are affected.** The error grows larger in the final overs where individual balls matter most for run rate calculations.

## Environment

- File: `application.py`
- Lines: training ~655-664, prediction ~997-1002
- Model: XGBoost via sklearn Pipeline

**Comment by anshul23102** [comment]
@Arnav-Singh-5080 I identified this feature mismatch while comparing the training pipeline with the prediction path in detail. The training phase computes `balls_left` using both the over number and the ball position within the over (`over * 6 + ball`), while the prediction path uses only whole overs (`overs * 6`). This means the model always receives the feature as if it is the very first ball of an over, which compounds into incorrect `crr` and `rrr` values.

I have a clear fix ready: add a "Balls in Current Over" slider (0-5) and wire it into the `balls_left` and `crr` derivations to match training-time feature engineering exactly. Would love to work on this.

assign/ 
GSSoC'26

**Comment by anshul23102** [comment]
Hi @Arnav-Singh-5080, I was assigned this issue. PR #184 is ready for review.

Added a "Balls in Current Over" slider (0-5) between the overs and wickets inputs. All three derived metrics (`balls_left`, `crr`, `rrr`) now use `overs * 6 + balls_in_over` for exact per-ball accuracy. Issue #156 is also covered by the same PR since the `max(120 - total_balls_bowled, 0)` guard correctly handles the over-20 boundary.
